File: backend/app/android_profile.py
"""Server-side builder for the strongSwan Android ".sswan" profile.

Why this exists as well as frontend/src/androidProfile.ts: the client-side
builder hands the whole profile to the external browser inside the URL
#fragment, and the embedded PKCS#12 makes that ~6.5 KB. Some Android
browsers/ROMs silently truncate long URLs, so the receiving page rebuilt a
CUT-OFF file and strongSwan rejected it with "Unterminated string at
character NNNN" (confirmed on a real device). Serving the profile from a
short token URL removes the length limit entirely — same fix already applied
to the iOS .mobileconfig path.

Keep in sync with frontend/src/androidProfile.ts (kept as the desktop
download fallback). Format verified against strongSwan's own docs:
docs.strongswan.org/docs/latest/os/androidVpnClientProfiles.html
"""

import json
import uuid


def build_sswan(endpoint_host: str, cert_base64: str, remark: str) -> str:
    return json.dumps(
        {
            "uuid": str(uuid.uuid4()),
            "name": remark,
            "type": "ikev2-cert",
            "remote": {"addr": endpoint_host},
            "local": {"p12": cert_base64, "rsa-pss": True},
            "ike-proposal": "aes256-sha256-modp2048",
            "esp-proposal": "aes128gcm16",
        },
        indent=2,
        ensure_ascii=False,
    )

File: frontend/src/androidProfile.ts
// Builds a strongSwan Android ".sswan" VPN Client Configuration profile —
// used instead of Amnezia's own app for the same reason as iOS's
// .mobileconfig: Amnezia's own source (isSupportedByCurrentPlatform)
// returns false for the Ipsec container on Android too, so its vpn://
// import would never work here either.
//
// strongSwan VPN Client (org.strongswan.android, free, official, Play
// Store/F-Droid) is a dedicated, actively maintained IKEv2 client that
// imports this exact JSON format — verified against strongSwan's own docs
// (docs.strongswan.org/docs/latest/os/androidVpnClientProfiles.html), not
// guessed, and matches the shape of Amnezia's own reference template
// (server_scripts/ipsec/strongswan.profile) field-for-field. File
// extension `.sswan`, media type `application/vnd.strongswan.profile`.
//
// No CA cert field needed: `pk12util`'s export (see backend/app/
// ikev2_manager.py) already bundles the signing CA into the client's
// PKCS#12 chain, which strongSwan extracts trust from directly — same
// reason Amnezia's own template omits `remote.cert` too.

import type { PeerCert, ServerParams } from "./types";

export function buildSswanProfile(server: ServerParams, cert: PeerCert, remark: string): string {
  return JSON.stringify(
    {
      uuid: crypto.randomUUID(),
      name: remark,
      type: "ikev2-cert",
      remote: { addr: server.endpoint_host },
      local: { p12: cert.cert_base64, "rsa-pss": true },
      "ike-proposal": "aes256-sha256-modp2048",
      "esp-proposal": "aes128gcm16",
    },
    null,
    2,
  );
}
